ocrd_network/ocrd_network/processor_server.py

```diff
@@ -0,0 +1,205 @@
+            processor_name=self.processor_name,
+            state=StateEnum.queued
+        )
+        await job.insert()
+
+        execution_failed = False
+        start_time = datetime.now()
+        await db_update_processing_job(
+            job_id=job_id,
+            state=StateEnum.running,
+            start_time=start_time
+        )
+        try:
+            background_tasks.add_task(
+                run_single_execution,
+                ProcessorClass=self.ProcessorClass,
+                executable=self.processor_name,
+                abs_path_to_mets=job.path_to_mets,
+                input_file_grps=job.input_file_grps,
+                output_file_grps=job.output_file_grps,
+                page_id=job.page_id,
+                parameters=job.parameters
+            )
+        except Exception as error:
+            self.log.debug(f"processor_name: {self.processor_name}, path_to_mets: {job.path_to_mets}, "
+                           f"input_grps: {job.input_file_grps}, output_file_grps: {job.output_file_grps}, "
+                           f"page_id: {job.page_id}, parameters: {job.parameters}")
+            self.log.error(error)
+            execution_failed = True
+        end_time = datetime.now()
+        exec_duration = calculate_execution_time(start_time, end_time)
+        job_state = StateEnum.success if not execution_failed else StateEnum.failed
+        await db_update_processing_job(
+            job_id=job_id,
+            state=job_state,
+            end_time=end_time,
+            exec_time=f'{exec_duration} ms'
+        )
+        return job.to_job_output()
+
+    def get_ocrd_tool(self):
+        if self.ocrd_tool:
+            return self.ocrd_tool
+        if self.ProcessorClass:
+            ocrd_tool = self.ProcessorClass(workspace=None, version=True).ocrd_tool
+        else:
+            ocrd_tool = get_ocrd_tool_json(self.processor_name)
+        return ocrd_tool
+
+    def get_version(self) -> str:
+        if self.version:
+            return self.version
+        if self.ProcessorClass:
+            version_str = self.ProcessorClass(workspace=None, version=True).version
+        else:
+            version_str = run(
+                [self.processor_name, '--version'],
+                stdout=PIPE,
+                check=True,
+                universal_newlines=True
+            ).stdout
+        return version_str
+
+    def run_server(self, host, port, access_log=False):
+        # TODO: Provide more flexibility for configuring file logging (i.e. via ENV variables)
+        file_handler = logging.FileHandler(f'/tmp/server_{self.processor_name}_{port}_{getpid()}.log', mode='a')
+        logging_format = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+        file_handler.setFormatter(logging.Formatter(logging_format))
+        file_handler.setLevel(logging.DEBUG)
+        self.log.addHandler(file_handler)
+        uvicorn.run(self, host=host, port=port, access_log=access_log)
```

Comment:
I understand, uvicorn (as does gunicorn) does not make it easy to deploy the server programmatically. But we should not dismiss the multiprocessing/worker model outright just because of that.

In https://github.com/OCR-D/core/pull/652, I also [used the CLI to create an app instance with runtime parameters](https://github.com/bertsky/core/blob/d98daa80695bd4bd5b0f23d4435a3abb0d524f54/ocrd/ocrd/decorators/__init__.py#L51-L67), and had my server simply [inherit from gunicorn.app.base.BaseApplication](https://github.com/bertsky/core/blob/d98daa80695bd4bd5b0f23d4435a3abb0d524f54/ocrd/ocrd/server.py#L18-L69) so I could run it – including the preforking multi-worker paradigm.

Here, since we also have to accommodate FastAPI, we would have to do some additional steps:
- use the [same inheritance pattern](https://docs.gunicorn.org/en/stable/custom.html#custom-application) for the ProcessingServer here
- tell [gunicorn to use uvicorn as worker class](https://www.uvicorn.org/#running-with-gunicorn)
- instead of the cmdline string, pass [`worker_class` as runtime option](https://github.com/benoitc/gunicorn/blob/792edf6d9aabcbfb84e76be1d722ac49c32dc027/gunicorn/config.py#L656-L657), pointing to the imported [`uvicorn.workers.UvicornWorker`](https://github.com/encode/uvicorn/blob/16d5101433bbfacf18e6ce9d21041164995b8bbb/uvicorn/workers.py#L14)

Reply:
> I understand, uvicorn (as does gunicorn) does not make it easy to deploy the server programmatically. But we should not dismiss the multiprocessing/worker model outright just because of that.

Of course not! Let's have a discussion and a clear vision of how we want to implement that so that all of us agree on common ground. @tdoan2010, please join the discussion.